x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/TransformProgressGatherer.java

```diff
@@ -0,0 +1,114 @@
+        MultiSearchRequestBuilder request = client.prepareMultiSearch()
+            .add(client.prepareSearch(config.getSource().getIndex())
+                .setSize(0)
+                .setTrackTotalHits(true)
+                .setQuery(config.getSource().getQueryConfig().getQuery())
+                .request());
+
+        if (cursor != null) {
+            request.add(client.prepareSearch(config.getSource().getIndex())
+                .setSize(0)
+                .setTrackTotalHits(true)
+                .setQuery(docsLeftQuery(config, cursor))
+                .request());
+        }
+        ActionListener<MultiSearchResponse> multiSearchResponseActionListener = ActionListener.wrap(
+            mSearchResponse -> {
+                long[] counts = new long[]{0, 0};
+                for (int i = 0; i < mSearchResponse.getResponses().length; i++) {
+                    MultiSearchResponse.Item item = mSearchResponse.getResponses()[i];
+                    if (item.isFailure()) {
+                        progressListener.onFailure(new ElasticsearchException("Failed gather transform task progress", item.getFailure()));
+                        return;
+                    }
+                    counts[i] = item.getResponse().getHits().getTotalHits().value;
+                }
+                // It stands to reason that total document count will be greater than the remaining document count as the
+                // remaining document count query contains the user provided query as a filter, thus should at most have the same
+                // amount of documents as the total document count
+                // If the mSearchResponse only contains 1 response, that response is the total document count
+                long totalDocs = Math.max(counts[0], counts[1]);
+                long remainingDocs = mSearchResponse.getResponses().length == 1 ? totalDocs : Math.min(counts[0], counts[1]);
+                progressListener.onResponse(new DataFrameTransformProgress(totalDocs, remainingDocs));
+            },
+            progressListener::onFailure
+        );
+        ClientHelper.executeAsyncWithOrigin(client.threadPool().getThreadContext(),
+            ClientHelper.DATA_FRAME_ORIGIN,
+            request.request(),
+            multiSearchResponseActionListener, client::multiSearch);
+    }
+
+    private static QueryBuilder docsLeftQuery(DataFrameTransformConfig config, Map<String, Object> cursor) {
+        if (cursor == null) {
+            return config.getSource().getQueryConfig().getQuery();
+        }
+        Map<String, SingleGroupSource<?>> groupConfig = config.getPivotConfig().getGroupConfig().getGroups();
+        List<Map.Entry<String, Object>> searchedGroupings = new ArrayList<>(cursor.size());
+        BoolQueryBuilder shouldQueries = QueryBuilders.boolQuery();
+        for (Map.Entry<String, Object> grouping : cursor.entrySet()) {
```

Comment:
Is the iteration order important when building this query?

Reply:
I drew a picture the answer is no

Reply:
The iteration order is important as the first item indicates the first thing we should sort by and then each subsequent grouping is added to restrict the bucket in which the rest of the keys are searched.

This comment points out a pretty bad bug :(. Apparently we sort the cursor by key, and thus breaks this completely
https://github.com/elastic/elasticsearch/blob/f83f72cf69ab92b8de5543b383944b340988b97d/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/dataframe/transforms/DataFrameTransformState.java#L94

I will look to see why we are doing a treehashmap there instead of keeping the order provided by composite agg and if the composite agg order is guaruteed. If not, the loop will have to be flipped and this will have to iterate over the `groupConfig` instead of the cursor

Reply:
Looking at composite agg results
```
"after_key" : {
        "origin" : "Abu Dhabi International Airport",
        "dest" : "Ataturk International Airport",
        "weekly" : 1557100800000
      },
```
And
https://github.com/elastic/elasticsearch/blob/c26d4ac6c136c32b2f447a974da37a618927561b/server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/ParsedComposite.java#L37

Indicates that the order of the cursor matches the order of the sources. So, we are breaking that order when we use a `TreeHashMap`. @hendrikmuhs Was there a particular reason in using a `TreeHashMap` any objections using a `LinkedHashMap`?

Reply:
@hendrikmuhs latest commit adjusts this https://github.com/elastic/elasticsearch/pull/41278/commits/5ac8e5b8441d07a8b3ae0ac3d02889d27b09cfe9

Let me know if this is acceptable

Reply:
`TreeMap` is what rollup uses, but I think there is no technical reason for it, composite agg does not seem to require order, so we could even use a map.